Fix account link no internet on startup

wmoss opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Currently when the internet is not available (or it fails to fetch the list of cloud services for some other reason) the fetch for cloud services in `account_link` just silently returns `[]`. This causes downstream integrations using `account_link` for ouath configurations to get weird errors about there being no implementations instead of something more informative and actionable. All integrations (except August and Yale, that were just fixed) currently throw an unrecoverable error when this happens, which means they need to be manually reconfigured when internet returns, instead of coming back up automatically like they should. There are more details in this issue, https://github.com/home-assistant/core/issues/153956.

~~The three commits in this PR,~~
1. ~~Change the oauth configuration implementation that `account_link` provides to throw an informative error message when it can't fetch the services.~~
2. ~~Use the new error message to fix the miele integration~~
3. ~~Use the new error message to fix the yolink integration
I'm happy to split this out into three PRs, it just felt easier to explain why I wanted to make the change in (1) with the examples in (2) and (3) and the code changes are small enough that it didn't feel too bad. I realized this is not perfect practice, but I didn't see people using stacked PRs, which is how I'd normally handle this.~~

~~Note: Since `ImplementationUnavailableError` inherits from `ValueError` the recent fixes to Yale and August will still work (although I, obviously, this they should switch to catching this error to avoid accidentally catching some other `ValueError`)~~

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #153956
- This PR is related to issue: #154244 and #154245
- Link to documentation pull request: 
- Link to developer documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/2850
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x]  The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr] (see note above)
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

wmoss committed:
Have account_link oauth configuration throw an error when services can't be fetched

Currently when the internet is not available (or it fails to fetch the list of cloud services for some other reason) the fetch for cloud services silently returns `[]`. This causes downstream integration configurations to get weird errors about there being no implementations instead of something more informative and actionable. All integreations (except August and Yale, that were just fixed) currently throw an unrecoverable error when this happens, which means they need to be manually reconfigured when internet returns, instead of coming back up automatically.

astrandb reviewed (commented):
Should the changes in miele and yolink really be included in this PR??
There are >50 integrations using `async_get_config_entry_implementation()` without try/except and all of these are likely suffering from the same problem.
I don't know what the best approach is - to ask the codeowners to update their integrations as in these examples or to create some sort of upstream fix - or both.

astrandb commented on homeassistant/components/miele/__init__.py:
```diff
@@ -44,7 +45,10 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    try:
+        implementation = await async_get_config_entry_implementation(hass, entry)
+    except ImplementationUnavailableError as err:
+        raise ConfigEntryNotReady("OAuth implementation unavailable") from err
```
The message should be translated. See the raise statements a couple of lines below.

wmoss commented:
@astrandb I am totally happy to pull the integration changes out of this PR into separate ones--as I noted in the description they are in here more as an explanation for why I want to add the different exception handling because I agree having them be separate is easier for review otherwise.

As to a more general solution, I couldn't figure one out, but am very open to suggestions. `account_link` gets setup as part of the `cloud` component, but it's only part of it, so I don't think you want to check the services call during that setup and fail the `cloud` component entirely if it fails (I did test it out of curiosity and it also doesn't stop the integrations from doing their configuration and then failing anyways).

balloob commented on homeassistant/components/cloud/account_link.py:
```diff
@@ -39,6 +39,10 @@ async def async_provide_implementation(
     """Provide an implementation for a domain."""
     services = await _get_services(hass)
+    if services is None:
+        raise config_entry_oauth2_flow.ImplementationUnavailableError(
```
The problem with this is that if account linking would ever be down, it would also prevent local implementations from not being able to load.

balloob replied:
If a config entry is configured to load the local one, it now does:

* local implementation -> return local one
* Ask cloud -> exception

And gets an exception back, instead of the local implementation.

balloob replied:
Ah nevermind, looked closer, should be fine.

balloob commented on homeassistant/helpers/config_entry_oauth2_flow.py:
```diff
@@ -61,6 +61,10 @@
 OAUTH_TOKEN_TIMEOUT_SEC = 30
 
 
+class ImplementationUnavailableError(ValueError):
```
let's inherit from HomeAssistantError

balloob reviewed (dismissed):
looks good. 1 comment and please add a test when we have 2 implementation providers, and the one that is not requested raises.

MartinHjelmare commented on homeassistant/components/cloud/account_link.py:
```diff
@@ -72,7 +76,7 @@ async def _get_services(hass: HomeAssistant) -> list[dict[str, Any]]:
         )
     except (aiohttp.ClientError, TimeoutError):
-        return []
+        return None
```
I think we should raise an exception here instead. It's weird to catch an exception, return None, and then raise an exception.

We could simply raise the `ImplementationUnavailableError` already here.

wmoss commented on homeassistant/helpers/config_entry_oauth2_flow.py:
```diff
@@ -563,9 +568,16 @@ async def async_get_implementations(
+            exceptions.append(err)
+
+    if len(registered) == 0 and len(exceptions) > 0:
+        raise ImplementationUnavailableError(*exceptions)
```
This code is new since the previous review and handles the case @balloob called out around local succeeding and cloud failing.

wmoss commented:
Thanks for the detailed reviews @balloob and @MartinHjelmare!

I've made the changes you requested and this is ready for review again. Changes of note,
1. Removed the integration fixes. I'll push those up separately.
2. Changed the logic to handle local succeeding and cloud failing correctly (e.g. return the local one and don't `raise`). I also added a few tests around this case to make sure it's working correctly.

MartinHjelmare commented:
For the future, please don't amend or squash commits after review has started to make it easier for readers to track changes. Thanks!

wmoss commented:
@MartinHjelmare @balloob any chance you can give this another look? Thanks!

MartinHjelmare commented on tests/helpers/test_config_entry_oauth2_flow.py:
```diff
@@ -1137,3 +1137,120 @@ def test_compute_code_challenge_invalid_code_verifier(code_verifier: str) -> Non
+async def test_async_get_config_entry_implementation_missing_provider(
+    hass: HomeAssistant,
+) -> None:
+    """Test async_get_config_entry_implementation when one provider fails and the other is empty."""
```
```suggestion
    """Test async_get_config_entry_implementation when both providers are empty."""
```

MartinHjelmare commented on homeassistant/components/cloud/account_link.py:
```diff
@@ -71,8 +71,10 @@ async def _get_services(hass: HomeAssistant) -> list[dict[str, Any]]:
-        return []
+    except (aiohttp.ClientError, TimeoutError) as err:
+        raise config_entry_oauth2_flow.ImplementationUnavailableError(
+            "Cannot provide OAuth2 implementation for cloud services. Failed to fetch from account link server."
```
```suggestion
            "Cannot provide OAuth2 implementation for cloud services. "
            "Failed to fetch from account link server."
```

wmoss commented on tests/helpers/test_config_entry_oauth2_flow.py:
```diff
@@ -1137,3 +1137,120 @@ def test_compute_code_challenge_invalid_code_verifier(code_verifier: str) -> Non
+        },
+    )
+
+    # This should fail since the local provider returned an empty list and the cloud provider raised an exception
```
It would be great to encode these line length restrictions in style rules.

wmoss committed:
Addressing comments from @MartinHjelmare

MartinHjelmare commented on tests/helpers/test_config_entry_oauth2_flow.py:
```diff
@@ -1137,3 +1137,116 @@ def test_compute_code_challenge_invalid_code_verifier(code_verifier: str) -> Non
+    )
+
+    # This should succeed and return the local implementation
+    # even though the failing cloud provider raised an exception
```
```suggestion
    # even though the failing cloud provider raised an exception.
```

MartinHjelmare commented on tests/helpers/test_config_entry_oauth2_flow.py:
```diff
@@ -1137,3 +1137,116 @@ def test_compute_code_challenge_invalid_code_verifier(code_verifier: str) -> Non
+        },
+    )
+
+    # This should fail since both providers are empty
```
```suggestion
    # This should fail since both providers are empty.
```

MartinHjelmare committed:
Add period

MartinHjelmare committed:
Add period

MartinHjelmare committed:
Add period

MartinHjelmare reviewed (approved):
Thanks!

wmoss committed:
Update the scaffold scripts for the new error handling

ludeeus commented:
Should this not be marked as a breaking change?

As several core integrations needed adjustments, should we assume custom integrations are affected as well?

MartinHjelmare commented:
I suggest we wait with merging this until after the beta cut on Wednesday. Then we have a whole release cycle to update integrations.

wmoss committed:
Merge branch 'dev' into fix-account_link-no-internet-on-startup

MartinHjelmare committed:
Merge branch 'dev' into fix-account_link-no-internet-on-startup
